src/main/java/com/gradle/CustomBuildScanEnhancements.java

```diff
@@ -55,215 +71,294 @@ void apply() {
-            if (Stream.of(pipelineName, pipelineNumber, stageName, stageNumber, jobName, goServerUrl).allMatch(Optional::isPresent)) {
-                //noinspection OptionalGetWithoutIsPresent
-                String buildUrl = String.format("%s/tab/build/detail/%s/%s/%s/%s/%s",
-                    goServerUrl.get(), pipelineName.get(),
-                    pipelineNumber.get(), stageName.get(), stageNumber.get(), jobName.get());
-                buildScan.link("GoCD build", buildUrl);
-            } else if (goServerUrl.isPresent()) {
-                buildScan.link("GoCD", goServerUrl.get());
+        @Override
+        public void execute(BuildResult buildResult) {
+            if (CiUtils.isJenkins(providers) || CiUtils.isHudson(providers)) {
+                Optional<String> buildUrl = envVariable("BUILD_URL", providers);
+                Optional<String> buildNumber = envVariable("BUILD_NUMBER", providers);
+                Optional<String> nodeName = envVariable("NODE_NAME", providers);
+                Optional<String> jobName = envVariable("JOB_NAME", providers);
+                Optional<String> stageName = envVariable("STAGE_NAME", providers);
+
+                buildUrl.ifPresent(url ->
+                        buildScan.link(CiUtils.isJenkins(providers) ? "Jenkins build" : "Hudson build", url));
+                buildNumber.ifPresent(value ->
+                        buildScan.value("CI build number", value));
+                nodeName.ifPresent(value ->
+                        customValueSearchLinker.addCustomValueAndSearchLink("CI node", value));
+                jobName.ifPresent(value ->
+                        customValueSearchLinker.addCustomValueAndSearchLink("CI job", value));
+                stageName.ifPresent(value ->
+                        customValueSearchLinker.addCustomValueAndSearchLink("CI stage", value));
+
+                jobName.ifPresent(j -> buildNumber.ifPresent(b -> {
+                    Map<String, String> params = new LinkedHashMap<>();
+                    params.put("CI job", j);
+                    params.put("CI build number", b);
+                    customValueSearchLinker.registerLink("CI pipeline", params);
+                }));
+            }
+
+            if (CiUtils.isTeamCity(providers)) {
+                Optional<String> teamcityBuildPropertiesFile = envVariable("TEAMCITY_BUILD_PROPERTIES_FILE", providers);
+                if (teamcityBuildPropertiesFile.isPresent()) {
+                    Properties buildProperties = readPropertiesFile(teamcityBuildPropertiesFile.get(), providers, projectDirectory.get());
+
+                    String teamcityConfigFile = buildProperties.getProperty("teamcity.configuration.properties.file");
+                    if (isNotEmpty(teamcityConfigFile)) {
+                        Properties configProperties = readPropertiesFile(teamcityConfigFile, providers, projectDirectory.get());
```

Comment:
In the previous version we would only read the file if the build id was also present. why did we change this? I'm not saying either way is wrong or correct, I just want to understand.

Reply:
I found it easier to read and assumption is that build id is always present.
Let's revert to the original approach though

Reply:
I didn't mean to suggest anything, I was just curious if there was a reason behind it. :)